common/error.hpp

```diff
@@ -13,31 +13,42 @@
 //    limitations under the License.
 
 #include <string>
+#include <system_error>
 #include <type_traits>
 
 #ifndef MENDER_COMMON_ERROR_HPP
 #define MENDER_COMMON_ERROR_HPP
 
 namespace mender::common::error {
 
-template <typename ErrorCodeType>
+template <typename ErrorType>
 class Error {
 public:
-	static_assert(
-		std::is_enum<ErrorCodeType>::value, "Error requires an enum type for error codes");
-	ErrorCodeType error_code;
+	ErrorType error;
 	std::string message;
 
-	Error(const ErrorCodeType &ec, const std::string &msg) {
-		this->error_code = ec;
-		this->message = msg;
-	};
-	Error(const Error &e) {
-		this->error_code = e.error_code;
-		this->message = e.message;
-	};
+	Error() {
+	}
+	Error(const ErrorType &ec, const std::string &msg) :
+		error(ec),
+		message(msg) {
+	}
+	Error(const Error &e) :
+		error(e.error),
+		message(e.message) {
+	}
+
+	bool operator==(const Error &other) const {
+		return this->message == other.message && this->error == other.error;
+	}
+
+	operator bool() const {
+		return static_cast<bool>(this->error);
```

Comment:
How will this works with an `enum` error code?

Reply:
This assumes that an error code of zero means success, but admittedly, that is not very clear from this code. And although I do think it's the case for most error types, it is an assumption that could be wrong.

Question is: Is there a better way to solve it? We need some way to test whether the instance is an error or not...

Reply:
> Question is: Is there a better way to solve it? We need some way to test whether the instance is an error or not...

As discussed on Slack, it probably makes most sense to use `error_condition` with our custom categories and error codes and respect the rule there that error code 0 always means no error.